### PR TITLE
Pipeline: Fix crash for `random_pipeline_model_task_construct_failpoint`

### LINKED ISSUE #7361: ASAN fail: ComputeServerRunner.multipleQuery
https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/tiflash-sanitizer-daily/runs/1048/nodes/75/steps/79/log/?start=0

```
[2023-04-24T05:17:37.326Z] ==23007==ERROR: AddressSanitizer: heap-use-after-free on address 0x611002863348 at pc 0x00001e210a6a bp 0x7f2bb3290960 sp 0x7f2bb3290958
[2023-04-24T05:17:37.326Z] READ of size 1 at 0x611002863348 thread T64 (WaitReactor)
[2023-04-24T05:17:37.326Z] [92m[2023/04/24 13:17:33.129 +08:00] [0m[1m[90m[TRACE] [0m[35m[MPPTunnel.cpp:121] [0m["destructed tunnel obj!"] [0m[90m[source="MPP<query:<query_ts:0, local_query_id:13, server_id:1, start_ts:13>,task_id:2> tunnel2+-1"] [thread_id=3059][0m
[2023-04-24T05:17:37.326Z] [92m[2023/04/24 13:17:33.129 +08:00] [0m[1m[90m[DEBUG] [0m[35m[MemoryTracker.cpp:70] [0m["Peak memory usage (for query): 320.00 KiB."] [0m[90m[source=MemoryTracker] [thread_id=3059][0m
[2023-04-24T05:17:37.326Z] [92m[2023/04/24 13:17:33.155 +08:00] [0m[1m[90m[TRACE] [0m[35m[ExchangeReceiverSourceOp.cpp:78] [0m["recv 0 rows from exchange receiver for tunnel11+7, total recv row num: 0"] [0m[90m[source="MPP<query:<query_ts:0, local_query_id:14, server_id:1, start_ts:14>,task_id:7> ExchangeReceiver exchange_receiver_12"] [thread_id=44][0m
[2023-04-24T05:17:37.326Z] [92m[2023/04/24 13:17:33.155 +08:00] [0m[1m[90m[TRACE] [0m[35m[Task.cpp:102] [0m["switch status: RUNNING --> WAITING"] [0m[90m[source="MPP<query:<query_ts:0, local_query_id:14, server_id:1, start_ts:14>,task_id:7> 2"] [thread_id=44][0m
[2023-04-24T05:17:37.326Z] [92m[2023/04/24 13:17:33.843 +08:00] [0m[1m[94m[INFO] [0m[35m[FlashService.cpp:366] [0m["Handling establish mpp connection request: sender_meta {\n  start_ts: 13\n  task_id: 6\n  partition_id: 1\n  address: \"0.0.0.0:4932\"\n  local_query_id: 13\n  server_id: 1\n}\nreceiver_meta {\n  start_ts: 13\n  task_id: 1\n  address: \"0.0.0.0:4931\"\n  local_query_id: 13\n  server_id: 1\n}\n"] [0m[90m[source=FlashService] [thread_id=3180][0m
[2023-04-24T05:17:37.326Z]     #0 0x1e210a69 in std::__1::__optional_storage_base<std::__1::shared_ptr<DB::MPMCQueue<DB::Block> >, false>::has_value() const /usr/local/bin/../include/c++/v1/optional:296:22
[2023-04-24T05:17:37.326Z]     #1 0x1e210a69 in DB::PipelineExecutorStatus::onEventFinish() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Executor/PipelineExecutorStatus.cpp:110:40
[2023-04-24T05:17:37.326Z]     #2 0x1e25cbde in DB::Event::finish() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Pipeline/Schedule/Events/Event.cpp:162:17
[2023-04-24T05:17:37.326Z]     #3 0x1e25d972 in DB::Event::onTaskFinish() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Pipeline/Schedule/Events/Event.cpp:187:9
[2023-04-24T05:17:37.326Z]     #4 0x1e269c5d in DB::EventTask::~EventTask() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Pipeline/Schedule/Tasks/EventTask.cpp:49:12
[2023-04-24T05:17:37.326Z]     #5 0x1e27062f in DB::PipelineTask::~PipelineTask() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Pipeline/Schedule/Tasks/PipelineTask.h:22:7
[2023-04-24T05:17:37.326Z]     #6 0x1e27062f in DB::PipelineTask::~PipelineTask() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Pipeline/Schedule/Tasks/PipelineTask.h:22:7
[2023-04-24T05:17:37.326Z]     #7 0x1e250f03 in std::__1::default_delete<DB::Task>::operator()(DB::Task*) const /usr/local/bin/../include/c++/v1/__memory/unique_ptr.h:57:5
[2023-04-24T05:17:37.326Z]     #8 0x1e250f03 in std::__1::unique_ptr<DB::Task, std::__1::default_delete<DB::Task> >::reset(DB::Task*) /usr/local/bin/../include/c++/v1/__memory/unique_ptr.h:318:7
[2023-04-24T05:17:37.327Z]     #9 0x1e250f03 in DB::(anonymous namespace)::Spinner::awaitAndPushReadyTask(std::__1::unique_ptr<DB::Task, std::__1::default_delete<DB::Task> >&&) /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Pipeline/Schedule/WaitReactor.cpp:55:18
[2023-04-24T05:17:37.327Z]     #10 0x1e250f03 in DB::WaitReactor::loop() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Pipeline/Schedule/WaitReactor.cpp:157:25
[2023-04-24T05:17:37.327Z]     #11 0x1e257a43 in decltype(*(static_cast<DB::WaitReactor*>(fp0)).*fp()) std::__1::__invoke<void (DB::WaitReactor::*)() noexcept, DB::WaitReactor*, void>(void (DB::WaitReactor::*&&)() noexcept, DB::WaitReactor*&&) /usr/local/bin/../include/c++/v1/type_traits:3859:1
[2023-04-24T05:17:37.327Z]     #12 0x1e257a43 in void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::WaitReactor::*)() noexcept, DB::WaitReactor*, 2ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::WaitReactor::*)() noexcept, DB::WaitReactor*>&, std::__1::__tuple_indices<2ul>) /usr/local/bin/../include/c++/v1/thread:280:5
[2023-04-24T05:17:37.327Z]     #13 0x1e257a43 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::WaitReactor::*)() noexcept, DB::WaitReactor*> >(void*) /usr/local/bin/../include/c++/v1/thread:291:5
[2023-04-24T05:17:37.327Z]     #14 0x7f2bd55a1ea4 in start_thread pthread_create.c
[2023-04-24T05:17:37.327Z]     #15 0x7f2bd50b4b0c in clone (/lib64/libc.so.6+0xfeb0c)
```

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #6518  close #7361

Problem Summary:

### What is changed and how it works?
- If task creation fails and one task has already been successfully created, the task that has been created will be destructed in `scheduleImpl`. This will trigger `onTaskFinish`, leading to unexpected problems.
- refine the code of `scheduleImpl` and use `Event::tasks` to hold the tasks that ready to be scheduled.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Pipeline/Schedule/Events/Event.cpp [hunk 1]
```diff
@@ -113,16 +120,29 @@ void Event::schedule() noexcept
...
+
+void Event::scheduleTasks() noexcept
+{
+    RUNTIME_ASSERT(status == EventStatus::SCHEDULED);
+    RUNTIME_ASSERT(tasks.size() == static_cast<size_t>(unfinished_tasks));
     if (!tasks.empty())
     {
-        scheduleTasks(tasks);
+        // If query has already been cancelled, we can skip scheduling tasks.
+        // And then tasks will be destroyed and call `onTaskFinish`.
```
Comment (ywqzzy) [inline dbms/src/Flash/Pipeline/Schedule/Events/Event.cpp:139]:
```suggestion
        // Then tasks will be destroyed and call `onTaskFinish`.
```

**Review by ywqzzy — APPROVED** [review]

File: dbms/src/Flash/Pipeline/Schedule/Events/Event.h [hunk 2]
```diff
@@ -95,6 +98,9 @@ class Event : public std::enable_shared_from_this<Event>
 
     std::atomic_int32_t unfinished_inputs{0};
 
+    // hold the tasks that ready to be scheduled.
+    std::vector<TaskPtr> tasks;
```
Comment (yibin87) [inline dbms/src/Flash/Pipeline/Schedule/Events/Event.h:102]: tasks won't be modified by mutiple threads?

Comment (SeaRise) [reply dbms/src/Flash/Pipeline/Schedule/Events/Event.h:102]: Yes, `tasks ` only used in `Event::schedule()`.

**Pull request merged**